Libraries/LibMarkdown/Table.cpp

```diff
@@ -0,0 +1,221 @@
+    for (auto& col : m_columns) {
+        if (!first)
+            builder.append('|');
+        first = false;
+        size_t width = col.relative_width * segment_width_proportion;
+        write_aligned(col.header, width, col.alignment);
+    }
+
+    builder.append("\n");
+    for (size_t i = 0; i < view_width; ++i)
+        builder.append('-');
+    builder.append("\n");
+
+    for (size_t i = 0; i < m_row_count; ++i) {
+        bool first = true;
+        for (auto& col : m_columns) {
+            if (col.rows.size() <= i)
+                continue;
+            auto& cell = col.rows[i];
+
+            if (!first)
+                builder.append('|');
+            first = false;
+
+            size_t width = col.relative_width * segment_width_proportion;
+            write_aligned(cell, width, col.alignment);
+        }
+        builder.append("\n");
+    }
+
+    return builder.to_string();
+}
+
+String Table::render_to_html() const
+{
+    StringBuilder builder;
+
+    builder.append("<table>");
+    builder.append("<thead>");
+    builder.append("<tr>");
+    for (auto& column : m_columns) {
+        builder.append("<th>");
+        builder.append(column.header.render_to_html());
+        builder.append("</th>");
+    }
+    builder.append("</tr>");
+    builder.append("</thead>");
+    builder.append("<tbody>");
+    for (size_t i = 0; i < m_row_count; ++i) {
+        builder.append("<tr>");
+        for (auto& column : m_columns) {
+            if (column.rows.size() > i) {
+                builder.append("<td>");
+                builder.append(column.rows[i].render_to_html());
+                builder.append("</td>");
+            }
+        }
+        builder.append("</tr>");
+    }
+    builder.append("</tbody>");
+    builder.append("</table>");
+
+    return builder.to_string();
+}
+
+OwnPtr<Table> Table::parse(Vector<StringView>::ConstIterator& lines)
+{
+    auto peek_it = lines;
+    auto first_line = *peek_it;
+    if (!first_line.starts_with('|'))
+        return nullptr;
+
+    ++peek_it;
+
+    if (peek_it.is_end())
+        return nullptr;
+
+    auto header_segments = first_line.split_view('|', true);
+    auto header_delimiters = peek_it->split_view('|', true);
+
+    header_segments.take_first();
+    header_delimiters.take_first();
+    if (header_segments.last().is_empty())
+        header_segments.take_last();
+    if (header_delimiters.last().is_empty())
+        header_delimiters.take_last();
+
+    ++peek_it;
+
+    if (header_delimiters.size() != header_segments.size())
+        return nullptr;
+
+    if (header_delimiters.is_empty())
+        return nullptr;
+
+    size_t total_width = 0;
+
+    auto table = make<Table>();
+    table->m_columns.resize(header_delimiters.size());
+
+    for (size_t i = 0; i < header_segments.size(); ++i) {
+        auto text_option = Text::parse(header_segments[i]);
+        ASSERT(text_option.has_value()); // Text::parse() cannot fail yet returns an Optional. why?
```

Comment:
That's right, it cannot fail, yet we pretend it can. I thought that was best, in case we ever add availability to it.

You're basically expected to treat this as a parsing failure rather than asserting. E.g. `return nullptr;` yourself.

Reply:
a parse failure in the header is fine, but what should a parse failure inside a structure that is _known_ to be a table do?
be ignored, treated as literal text, or just immediately fail the table parse?

|a|b|
|---|---|
|`|x|

github seems to take it as literal text